go/oasis-node/cmd/debug/txsource/workload/governance.go

```diff
@@ -0,0 +1,479 @@
+					// Proposal execution should fail.
+					shouldFail = true
+					break
+				}
+			}
+			if shouldFail {
+				continue
+			}
+			// Add the pending upgrade.
+			g.pendingUpgrades = append(g.pendingUpgrades, &ap.Content.Upgrade.Descriptor)
+		}
+		delete(g.ensureYesVote, ap.ID)
+	}
+	g.currentEpoch = epoch
+	g.activeProposals = remainingActiveProposals
+
+	return nil
+}
+
+// Implements Workload.
+func (g *governanceWorkload) NeedsFunds() bool {
+	return true
+}
+
+// Implements Workload.
+func (g *governanceWorkload) Run(
+	gracefulExit context.Context,
+	rng *rand.Rand,
+	conn *grpc.ClientConn,
+	cnsc consensus.ClientBackend,
+	sm consensus.SubmissionManager,
+	fundingAccount signature.Signer,
+	validatorEntities []signature.Signer,
+) error {
+	var err error
+
+	// Initialize state.
+	g.BaseWorkload.Init(cnsc, sm, fundingAccount)
+	g.rng = rng
+	g.ctx = context.Background()
+
+	g.parameters, err = cnsc.Governance().ConsensusParameters(g.ctx, consensus.HeightLatest)
+	if err != nil {
+		return fmt.Errorf("error querying governance consensus parameters: %w", err)
+	}
+
+	g.governance = cnsc.Governance()
+
+	g.validatorEntities = validatorEntities
+	if len(g.validatorEntities) == 0 {
+		return fmt.Errorf("workload requires validator entities")
+	}
+
+	g.proposerAccounts = make([]*struct {
+		signer  signature.Signer
+		address staking.Address
+		nonce   uint64
+	}, numProposerAccounts)
+	fac := memorySigner.NewFactory()
+	for i := range g.proposerAccounts {
+		var signer signature.Signer
+		signer, err = fac.Generate(signature.SignerEntity, rng)
+		if err != nil {
+			return fmt.Errorf("memory signer factory Generate account %d: %w", i, err)
+		}
+		g.proposerAccounts[i] = &struct {
+			signer  signature.Signer
+			address staking.Address
+			nonce   uint64
+		}{
+			signer:  signer,
+			address: staking.NewAddress(signer.Public()),
+		}
+		if err = g.TransferFunds(
+			g.ctx,
+			fundingAccount,
+			g.proposerAccounts[i].address,
+			g.parameters.MinProposalDeposit.ToBigInt().Uint64()*1000,
+		); err != nil {
+			return fmt.Errorf("account funding failure: %w", err)
+		}
+	}
+
+	g.ensureYesVote = make(map[uint64]bool)
+
+	// Main workload loop.
+	for {
+		select {
+		case <-time.After(1 * time.Second):
+		case <-gracefulExit.Done():
+			g.Logger.Debug("time's up")
+			return nil
+		}
+
+		var epoch epochtime.EpochTime
+		epoch, err = cnsc.GetEpoch(g.ctx, consensus.HeightLatest)
+		if err != nil {
+			return fmt.Errorf("querying epoch: %w", err)
+		}
+
+		// Ensure epoch transition won't happen for the next 10 blocks. Otherwise
+		// if the epoch transition happens during the loop iteration, we can have
+		// outdated local state.
+		var epoch2 epochtime.EpochTime
+		var block *consensus.Block
+		block, err = cnsc.GetBlock(g.ctx, consensus.HeightLatest)
+		if err != nil {
+			return fmt.Errorf("querying block: %w", err)
+		}
+		epoch2, err = cnsc.GetEpoch(g.ctx, block.Height+10)
```

Comment:
This will not work with the new beacon as `GetEpoch` there performs a query which will fail if you give it a future height. Also with the new beacon there is no way to know for sure when the next transition will happen.

Reply:
Hm yeah that's right. Ok i'll update the workload and turn things around a bit:
- first a request will be made (e.g. submit proposal, vote ...)
- after the request the workload will check if the request should have failed or succeed based on the state at that height

Reply:
actually simplified it a bit more even. In case the unexpected result happens (e.g. a proposal is successfully submitted when it shouldn't be - or otherwise) the workload checks if the epoch transition just happened, and in case it did it doesn't fail the whole workload and just emits an error.

Reply:
Yeah that sounds like the right solution, thanks!

Reply:
Np! thanks for noticing that this wouldn't work with the new beacon :grin: forgot about it.